src/main/java/com/mongodb/hibernate/internal/translate/AbstractMqlTranslator.java

```diff
@@ -161,6 +172,8 @@ abstract class AbstractMqlTranslator<T extends JdbcOperation> implements SqlAstT
 
     private final Set<String> affectedTableNames = new HashSet<>();
 
+    private final Set<Predicate> subPredicatesInNegatedGroupedPredicate = new HashSet<>();
```

Comment:
It's notable whenever we have to add a new field to this class in order to share state as we visit the HQL AST, as it creates opportunities for that state to get out of sync with reality.  In this case, for example, I see code which adds to the set, and code that queries for a predicate in the set, but no code that removes anything from the set. I'm not sure, but it seems like there is a bug lurking in here somewhere.

Reply:
My concern is not at all related to performance, but rather correctness. Is it possible that with a more complicated query the `subPredicatesInNegatedGroupedPredicate` field will need to be used more than once to pass information up the stack?  If so, it seems like it would need to be cleared after the value is transmitted, to prepare it to be used once again.  We were very careful with that in our use of `astVisitorValueHolder`, so I wonder if some of the same concerns apply to this field.

Reply:
I don't think we have such concern. Regardless of how complicated the query is, we have only one query as the translation process input, right? This set's scope will be bound to the translation process so bound to the query. Let us say the query contains multple levels of embedded subqueries (the most complicated query I could think of) and each subquery has its own negated grouped predicates. During the translation, each different `not (...)` will end up with different SQL AST nodes, so there is no conflict here from my understanding.

It is different from `astVisitorValueHolder` for it is a collection data structure, so different SQL AST nodes won't conflict and they simply be inserted and queried based on their SQL AST node reference.

Reply:
@vbabanin I'm going to leave this open for you to consider, but please feel free to resolve if you are satisified.